src/backend/columnar/columnar_customscan.c

```diff
@@ -631,15 +995,29 @@ ColumnarScanPath_PlanCustomPath(PlannerInfo *root,
 	 * CustomScan as the first field. Otherwise, copyObject() will fail to
 	 * copy the additional fields.
 	 */
-	CustomScan *cscan = (CustomScan *) newNode(sizeof(CustomScan), T_CustomScan);
+	CustomScan *cscan = makeNode(CustomScan);
 
 	cscan->methods = &ColumnarScanScanMethods;
 
-	/* Reduce RestrictInfo list to bare expressions; ignore pseudoconstants */
-	clauses = extract_actual_clauses(clauses, false);
+	/* XXX: also need to store projected column list for EXPLAIN */
+
+	if (EnableColumnarQualPushdown)
+	{
+		/*
+		 * List of pushed-down clauses. The Vars referencing other relations
+		 * will be changed into exec Params by create_customscan_plan().
+		 *
+		 * XXX: this just means what will be pushed into the columnar reader
+		 * code; some of these may not be usable. We should fix this by
+		 * passing down something more like ScanKeys, where we've already
+		 * verified that the operators match the btree opclass of the chunk
+		 * predicates.
+		 */
+		cscan->custom_exprs = copyObject(best_path->custom_private);
+	}
 
+	cscan->scan.plan.qual = extract_actual_clauses(clauses, false);
```

Comment:
similar to the comment I left for `AddColumnarScanPath`:

```suggestion
	/* extract actual clauses, not pseudo constants */
	bool pseudoconstant = false;
	cscan->scan.plan.qual = extract_actual_clauses(clauses, pseudoconstant);
```